### ISSUE TITLE
cat | cat | cat randomly fails

### ISSUE THREAD

**Issue by pjhades** [pjhades/rubbish#12]
To reproduce,

* run `cat | cat | cat`, sometimes its `cat | cat | cat | cat`
* type something
* hit Ctrl + C

You can find that `cat` does not display what you've typed on the terminal, and the shell crashes in the last step.

Or

* run `cat | cat | cat | cat`
* hit Ctrl + Z

Then the shell crashes.

**Comment by pjhades** [comment]
It's strange that after `cat | cat | cat | cat` is started, `ps` shows that some `cat` is in `T` state (stopped) rather than `S` state (sleep). That should be the cause of `cat` not displaying the input string.

**Comment by pjhades** [comment]
The stopped `cat` is possible if it is spawned too quickly so that `cat` starts reading from its standard input BEFORE we put it into foreground, in which case `cat` receives `SIGTTIN` and is stopped.

**Comment by pjhades** [comment]
Now, seriously, I think something must be wrong here:

```ruby
$n = 0
Signal.trap("SIGCHLD") { $n += 1 }

pid = Process.fork do
    # do nothing in the fucking child
end
p, s = Process.waitpid2 pid
puts "p=#{p}, s=#{s}, n=#{$n}"
```

Then we run it:

```bash
pluto @ misato: lab → ruby fuck.rb
p=7931, s=pid 7931 exit 0, n=0

pluto @ misato: lab → for i in $(seq 1 100); do ruby fuck.rb; done | grep 'n=1' | wc -l
1

pluto @ misato: lab → for i in $(seq 1 100); do ruby fuck.rb; done | grep 'n=1' | wc -l
1

pluto @ misato: lab → for i in $(seq 1 100); do ruby fuck.rb; done | grep 'n=1' | wc -l
6
```

I can't even ....

**Comment by pjhades** [comment]
The trick to avoid `SIGTTIN` is, once the first child in a job has been spawned, we can obtain the PGID, and just then we put that group into foreground. Then we `exec` the command in the child.

Bash is implemented in this way. See the `make_child()` and `execute_simple_command()` functions.

**Comment by pjhades** [comment]
Pay attention to `SIGTTOU`. `man tcsetpgrp` says

> If  `tcsetpgrp()`  is  called  by  a member of a background process group in its session, and the calling process is not blocking or ignoring `SIGTTOU`, a `SIGTTOU` signal is sent to all members of this background process group.

Thus **`SIGTTOU` should be ignored** by the children before they move themselves to foreground.

**Comment by pjhades** [comment]
Now the problem is the shell gets  `SIGCHLD` where there is no child exited.

**Comment by pjhades** [comment]
Could possibly be signal-related race condition, as the writings to `$n_sigchld` is not atomic.

Actually we may not need `$n_sigchld`. For the foreground jobs, we block on `wait`, everything's good. For the background jobs, we wait for them at certain places in a nonblocking manner. This may be enough to reap all the children without leaking any.

**Comment by pjhades** [comment]
Solved in ee3b9ab22afb7c237d8017bb59a3996b4ee380e2, YAY!
